focus-ios/Blockzilla/Utilities/SearchHistoryUtils.swift

```diff
@@ -27,29 +27,21 @@ class SearchHistoryUtils {
 
     static func pushSearchToStack(with searchedText: String) {
         // Check whether the lastSearch is the current search. If not, remove subsequent searches
-        if let lastSearch = currentStack.last, !lastSearch.isCurrentSearch {
-            for index in 0..<currentStack.count {
-                if currentStack[index].isCurrentSearch {
-                    currentStack.removeSubrange(index+1..<currentStack.count)
-                    break
-                }
-            }
+        if let lastIndex = currentStack.firstIndex(where: { !$0.isCurrentSearch }) {
+            currentStack.removeSubrange(lastIndex + 1..<currentStack.count)
```

Comment:
This change doesn't look correct. The stack should only be cleaned up if the lastSearch is not the current search.
```suggestion
        // Check whether the lastSearch is the current search. If not, remove subsequent searches
        if let lastSearch = currentStack.last,
           !lastSearch.isCurrentSearch,
           let lastIndex = currentStack.firstIndex(where: { !$0.isCurrentSearch }) {
            currentStack.removeSubrange(lastIndex + 1..<currentStack.count)
```

Reply:
Indeed, the change might not be correct. We want go and look for searches to be removed only if the last search is not the current search. Then we look for the first search that is current search and remove everything after it (not including it). After that we mark everything as not being current search and only afteeer that we add the new search and mark it accordingly. Ugh that was a lot. A similar implementation the might be more clear can be:

```swift
static func pushSearchToStack(with searchedText: String) {
    // Check if the stack is empty and directly append if it is
    guard !currentStack.isEmpty else {
        currentStack.append(textSearched(text: searchedText, isCurrentSearch: true))
        return
    }

    // Find the last `current search` and truncate the stack beyond this point
    if let lastIndex = currentStack.lastIndex(where: { $0.isCurrentSearch }) {
        currentStack = Array(currentStack.prefix(upTo: lastIndex + 1))
    }

    // Mark all as not current in a single pass and append new search
    currentStack = currentStack.map { textSearched(text: $0.text, isCurrentSearch: false) }
    currentStack.append(textSearched(text: searchedText, isCurrentSearch: true))
}
```

To show an example we start with
0: "apple"        (isCurrentSearch = false)
1: "banana"      (isCurrentSearch = false)
2: "cherry"       (isCurrentSearch = true)
3: "date"          (isCurrentSearch = false)

Then we truncate after index 2
0: "apple"        (isCurrentSearch = false)
1: "banana"      (isCurrentSearch = false)
2: "cherry"       (isCurrentSearch = true)

We set everything to false
0: "apple"        (isCurrentSearch = false)
1: "banana"      (isCurrentSearch = false)
2: "cherry"       (isCurrentSearch = false)

Then we add the search
0: "apple"           (isCurrentSearch = false)
1: "banana"        (isCurrentSearch = false)
2: "cherry"         (isCurrentSearch = false)
3: "elderberry"  (isCurrentSearch = true)